retworkx/visualization/matplotlib.py

```diff
@@ -0,0 +1,1153 @@
+
+        See `matplotlib.patches.ArrowStyle` for more options.
+
+    arrow_size : int (default=10)
+        For directed graphs, choose the size of the arrow head's length and
+        width. See `matplotlib.patches.FancyArrowPatch` for attribute
+        ``mutation_scale`` for more info.
+
+    node_size : scalar or array (default=300)
+        Size of nodes. Though the nodes are not drawn with this function, the
+        node size is used in determining edge positioning.
+
+    node_list : list, optional (default=graph.node_indexes())
+       This provides the node order for the `node_size` array (if it is an
+       array).
+
+    node_shape :  string (default='o')
+        The marker used for nodes, used in determining edge positioning.
+        Specification is as a `matplotlib.markers` marker, e.g. one of
+        'so^>v<dph8'.
+
+    label : None or string
+        Label for legend
+
+    min_source_margin : int (default=0)
+        The minimum margin (gap) at the begining of the edge at the source.
+
+    min_target_margin : int (default=0)
+        The minimum margin (gap) at the end of the edge at the target.
+
+    Returns
+    -------
+    list of matplotlib.patches.FancyArrowPatch
+        `FancyArrowPatch` instances of the directed edges
+
+    Notes
+    -----
+    For directed graphs, arrows are drawn at the head end.  Arrows can be
+    turned off with keyword arrows=False or by passing an arrowstyle without
+    an arrow on the end.
+
+    Be sure to include `node_size` as a keyword argument; arrows are
+    drawn considering the size of nodes.
+    """
+    try:
+        import matplotlib as mpl
+        import matplotlib.colors  # call as mpl.colors
+        import matplotlib.patches  # call as mpl.patches
+        import matplotlib.path  # call as mpl.path
+        import matplotlib.pyplot as plt
+    except ImportError as e:
+        raise ImportError(
+            "matplotlib needs to be installed prior to running "
+            "retworkx.visualization.mpl_draw(). You can install "
+            "matplotlib with:\n'pip install matplotlib'"
+        ) from e
+
+    if arrowstyle is None:
+        if isinstance(graph, retworkx.PyDiGraph) and arrows:
+            arrowstyle = "-|>"
+        else:
+            arrowstyle = "-"
+
+    if ax is None:
+        ax = plt.gca()
+
+    if edge_list is None:
+        edge_list = graph.edge_list()
+
+    if len(edge_list) == 0:  # no edges!
+        return []
+
+    if node_list is None:
+        node_list = list(graph.node_indexes())
+
+    # FancyArrowPatch handles color=None different from LineCollection
+    if edge_color is None:
+        edge_color = "k"
+
+    # set edge positions
+    edge_pos = np.asarray([(pos[e[0]], pos[e[1]]) for e in edge_list])
+
+    # Check if edge_color is an array of floats and map to edge_cmap.
+    # This is the only case handled differently from matplotlib
+    if (
+        np.iterable(edge_color)
+        and (len(edge_color) == len(edge_pos))
```

Comment:
Could it be possible for `edge_color` to be a dict? 
My reasoning is that we could have a default color and just change some of the nodes that we want to change using a dictionary with `edge_pos` as key values. Is that reasonable?

Reply:
Well `edge_pos` isn't user facing (it's just internal array of node positions for the nodes of an edge), but I think I see what your getting at here. The only way we can do different colors for a subset of edges now is multiple calls to `mpl_draw` with a subset of nodes and edges (the tests show examples of this). In networkx this kind of thing is handled normally by calling `draw_edges()` separately, but for this PR I opted for a single function for simplicity. This however causes some tension for more custom things like different colors for specific edges or nodes. I think there are 3 ways to address this:

1. Make the internal `draw_nodes`, `draw_edges`, etc functions part of the user facing api like networkx does.
2. Add support for dicts/mappings as an input to `edge_colors`, `node_colors`, `node_shape` (and maybe `node_size`) so that these properties can be set per node and per edge.
3. Add callback function support for `edge_colors`, `node_colors`, etc so that when we evaluate colors we can just call the user provided function to get a color string (or whatever other property.

None of these options are mutually exclusive so we can do any combination of them. How do you think we should proceed here?

Reply:
The most natural solution, at least for me, is the first one. However, exposing those functions may lead to some trouble. For the long term, I think the second solution can be better. What you think about doing both 1 and 2?

Reply:
I was on the fence about doing 1, but I agree it seems like it makes the most sense for user expectations. I think it'll be good as a followup (either before or after release). Same for 2, we can do that in a follow up. I'm going to merge this now (with your review) I'll open an issue about this after merging.